README.md
# node-red-puppeteer


### These node-red nodes expose puppeteer's API used for controlling Chrome/Chromium over the <a href='https://chromedevtools.github.io/devtools-protocol/'>DevTools Protocol</a>


Group | Node Name | Detailed information
------------- | ------------- | -------------
<b>Browser</b> | Launch 
<b>Browser</b> | New Page 
<b>Browser</b> | Close Browser 
<b>Page</b> | Click 
<b>Page</b> | Close 
<b>Page</b> | Content 
<b>Page</b> | Delete Cookies 
<b>Page</b> | Download 
<b>Page</b> | Focus 
<b>Page</b> | Get 
<b>Page</b> | Get Cookies 
<b>Page</b> | Go to 
<b>Page</b> | Screenshot 
<b>Page</b> | Viewport
<b>Page</b> | Set 
<b>Page</b> | Set LocalStorage
<b>Page</b> | Set Cookies 
<b>Page</b> | Upload 
<b>Page</b> | Wait for 
<b>Keyboard | Press 
<b>Keyboard | Type 
<b>Storage</b> | Rename 

#### Install 
```bash
npm install @robinck/node-red-puppeteer
```



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

## 1.0.1 - 2023-07-26
Updated README

## 1.0.0 - 2023-07-25
Initial release

### Added
- Changelog


puppeteer/page/click.js
module.exports = function (RED) {
  function PuppeteerPageClick(nodeConfig) {
    RED.nodes.createNode(this, nodeConfig);
    // Parse the click count to integer
    nodeConfig.clickCount = parseInt(nodeConfig.clickCount);
    var node = this; // Referencing the current node

    this.on("input", async function (msg, send, done) {
      try {
        // Parsing the selector from string input or from msg object
        let selector =
          nodeConfig.selectortype != "str"
            ? eval(nodeConfig.selectortype + "." + nodeConfig.selector)
            : nodeConfig.selector;
        // If the type of selector is set to flow or global, it needs to be parsed differently
        if (
          nodeConfig.selectortype == "flow" ||
          nodeConfig.selectortype == "global"
        ) {
          // Parsing the selector
          selector = this.context()[nodeConfig.selectortype].get(
            nodeConfig.selectortype
          );
        }

        // Waiting for the specified selector
        node.status({
          fill: "blue",
          shape: "ring",
          text: `Wait for ${selector}`,
        });
        await msg.puppeteer.page.waitForSelector(selector);

        // Clicking on the specified selector
        node.status({ fill: "blue", shape: "dot", text: `Click ${selector}` });
        await msg.puppeteer.page.click(selector, nodeConfig);

        // Selector clicked sucessfully
        node.status({
          fill: "green",
          shape: "dot",
          text: `Clicked ${selector}`,
        });
        // Sending the msg
        send(msg);

      } catch (e) {
        // If an error occured
        node.error(e);
        // Update the status
        node.status({ fill: "red", shape: "dot", text: e });
        // And update the message error property
        msg.error = e;
        send(msg);
      }

      // Clear status of the node
      setTimeout(() => {
        done();
        node.status({});
      }, (msg.error) ? 10000 : 3000);
    });
    this.on("close", function () {
      node.status({});
    });
    oneditprepare: function oneditprepare() {
      $("#node-input-clickCount").val(nodeConfig.clickCount);
      $("#node-input-delay").val(nodeConfig.delay);
      $("#node-input-button").val(nodeConfig.button);
      $("#node-input-name").val(nodeConfig.name);
    }
  }
  RED.nodes.registerType("puppeteer-page-click", PuppeteerPageClick);
};


puppeteer/page/rename.js
module.exports = function (RED) {
  function PuppeteerRename(nodeConfig) {
    RED.nodes.createNode(this, nodeConfig);
    var node = this; // Referencing the current node

    this.on("input", async function (msg, send, done) {
      try {
        // Parsing the oldFilePath from string input or from msg object
        let oldFilePath =
          nodeConfig.oldFilePathtype != "str"
            ? eval(nodeConfig.oldFilePathtype + "." + nodeConfig.oldFilePath)
            : nodeConfig.oldFilePath;
        // If the type of oldFilePath is set to flow or global, it needs to be parsed differently
        if (
          nodeConfig.oldFilePathtype == "flow" ||
          nodeConfig.oldFilePathtype == "global"
        ) {
          // Parsing the oldFilePath
          oldFilePath = this.context()[nodeConfig.oldFilePathtype].get(
            nodeConfig.oldFilePathtype
          );
        }

        // Parsing the newFilePath from string input or from msg object
        let newFilePath =
          nodeConfig.newFilePathtype == "msg"
            ? eval(nodeConfig.newFilePathtype + "." + nodeConfig.newFilePath)
            : nodeConfig.newFilePath;
        // If the type of newFilePath is set to flow or global, it needs to be parsed differently
        if (
          nodeConfig.newFilePathtype == "flow" ||
          nodeConfig.newFilePathtype == "global"
        ) {
          // Parsing the newFilePath
          newFilePath = this.context()[nodeConfig.newFilePathtype].get(
            nodeConfig.newFilePath
          );
        }

        if(oldFilePath && newFilePath && oldFilePath != '' && newFilePath != '') {
          const fs = require('fs');
          fs.renameSync(oldFilePath, newFilePath);
        } else {
          throw 'Please enter a valid file path';
        }

        // oldFilePath clicked sucessfully
        node.status({
          fill: "green",
          shape: "dot",
          text: `Clicked ${oldFilePath}`,
        });
        // Sending the msg
        send(msg);
      } catch (e) {
        // If an error occured
        node.error(e);
        // Update the status
        node.status({ fill: "red", shape: "dot", text: e });
        // And update the message error property
        msg.error = e;
        send(msg);
      }

      // Clear status of the node
      setTimeout(
        () => {
          done();
          node.status({});
        },
        msg.error ? 10000 : 3000
      );
    });
    this.on("close", function () {
      node.status({});
    });
    oneditprepare: function oneditprepare() {
      $("#node-input-oldFilePath").val(nodeConfig.oldFilePath);
      $("#node-input-newFilePath").val(nodeConfig.newFilePath);
    }
  }
  RED.nodes.registerType("puppeteer-rename", PuppeteerRename);
};


puppeteer/page/download.js
module.exports = function (RED) {
  function PuppeteerPageDownload(nodeConfig) {
    RED.nodes.createNode(this, nodeConfig);
    // Parse the click count to integer
    nodeConfig.clickCount = parseInt(nodeConfig.clickCount);
    var node = this; // Referencing the current node

    this.on("input", async function (msg, send, done) {
      try {
        // Parsing the selector from string input or from msg object
        let selector =
          nodeConfig.selectortype != "str"
            ? eval(nodeConfig.selectortype + "." + nodeConfig.selector)
            : nodeConfig.selector;
        // If the type of selector is set to flow or global, it needs to be parsed differently
        if (
          nodeConfig.selectortype == "flow" ||
          nodeConfig.selectortype == "global"
        ) {
          // Parsing the selector
          selector = this.context()[nodeConfig.selectortype].get(
            nodeConfig.selectortype
          );
        }

        // Parsing the downloadPath from string input or from msg object
        let downloadPath =
          nodeConfig.downloadPathtype == "msg"
            ? eval(nodeConfig.downloadPathtype + "." + nodeConfig.downloadPath)
            : nodeConfig.downloadPath;
        // If the type of downloadPath is set to flow or global, it needs to be parsed differently
        if (
          nodeConfig.downloadPathtype == "flow" ||
          nodeConfig.downloadPathtype == "global"
        ) {
          // Parsing the downloadPath
          downloadPath = this.context()[nodeConfig.downloadPathtype].get(
            nodeConfig.downloadPath
          );
        }

        // Parsing the fileName from string input or from msg object
        let fileName =
          nodeConfig.fileNametype == "msg"
            ? eval(nodeConfig.fileNametype + "." + nodeConfig.fileName)
            : nodeConfig.fileName;
        // If the type of fileName is set to flow or global, it needs to be parsed differently
        if (
          nodeConfig.fileNametype == "flow" ||
          nodeConfig.fileNametype == "global"
        ) {
          // Parsing the fileName
          fileName = this.context()[nodeConfig.fileNametype].get(
            nodeConfig.fileName
          );
        }

        // If download path is defined
        if (downloadPath && downloadPath != "") {
          // Enable requests interception
          await msg.puppeteer.page.setRequestInterception(true);

          // When request comes
          msg.puppeteer.page.on("request", (interceptedRequest) => {
            // And for everything that ends with zip (for now only supported file type)
            if (interceptedRequest.url().endsWith(".zip")) {
              interceptedRequest.continue({ url: "chrome://downloads/" }); // Continue to downloads url where download path is defined
            } else {
              interceptedRequest.continue(); // Continue the requets as usual
            }
          });

          if (fileName && fileName != "") {
            msg.puppeteer.page.on("response", async (response) => {
              // node.warn(response);
              const url = response.url();
              const headers = response.headers();
              const regex = /filename\*?=[^'']*''([^']*)/;

              if (headers["content-disposition"]) {
                const match = regex.exec(headers["content-disposition"]);
                node.warn(headers["content-disposition"]);
                if (match && match[1]) {
                  const fs = require("fs");
                  const downloadedFileName = match[1];
                  const extensionRegex = /\.[^.]*$/;
                  const extension = extensionRegex.exec(downloadedFileName);
                  if(extension) {
                    msg.old = `${downloadPath}/${downloadedFileName}`;
                    msg.new = `${downloadPath}/${fileName}${extension[0]}`;
                    node.send(msg);
                  }
                  // extension ? msg.old
                  //   : node.send({ payload: null });
                  // const fileExtension = extension ? fs.renameSync(`${downloadPath}/${downloadedFileName}`, `${downloadPath}/${fileName}.${extension[0]}`) : node.warn(`File has no extension! saved as ${downloadedFileName} in ${downloadPath}!`);
                  // fs.renameSync(`${downloadPath}/${downloadedFileName}`, `${downloadPath}/${fileName}.${extension}`);
                }
              }
            });
          }
        }

        // Waiting for the specified selector
        node.status({
          fill: "blue",
          shape: "ring",
          text: `Wait for ${selector}`,
        });
        await msg.puppeteer.page.waitForSelector(selector);

        // Clicking on the specified selector
        node.status({ fill: "blue", shape: "dot", text: `Click ${selector}` });
        await msg.puppeteer.page.click(selector, nodeConfig);

        // If download path was specified
        if (downloadPath && downloadPath != "") {
          // Set the download path and download the file to it
          await msg.puppeteer.page._client.send("Page.setDownloadBehavior", {
            behavior: "allow",
            downloadPath: downloadPath,
          });
        }

        // Selector clicked sucessfully
        node.status({
          fill: "green",
          shape: "dot",
          text: `Clicked ${selector}`,
        });
        if (!fileName || fileName == "")
          // Sending the msg
          send(msg);
      } catch (e) {
        // If an error occured
        node.error(e);
        // Update the status
        node.status({ fill: "red", shape: "dot", text: e });
        // And update the message error property
        msg.error = e;
        send(msg);
      }

      // Clear status of the node
      setTimeout(
        () => {
          done();
          node.status({});
        },
        msg.error ? 10000 : 3000
      );
    });
    this.on("close", function () {
      node.status({});
    });
    oneditprepare: function oneditprepare() {
      $("#node-input-clickCount").val(nodeConfig.clickCount);
      $("#node-input-delay").val(nodeConfig.delay);
      $("#node-input-button").val(nodeConfig.button);
      $("#node-input-name").val(nodeConfig.name);
      $("#node-input-downloadPath").val(nodeConfig.downloadPath);
    }
  }
  RED.nodes.registerType("puppeteer-page-download", PuppeteerPageDownload);
};


puppeteer/page/viewport.js
module.exports = function (RED) {
  function PuppeteerPageViewport(config) {
    RED.nodes.createNode(this, config);
    var node = this; // Referencing the current node

    this.on("input", async function (msg, send, done) {
      try {
        // Capturing screen
        node.status({
          fill: "blue",
          shape: "dot",
          text: `Change viewport ...`,
        });
        await msg.puppeteer.page.setViewport({
          width: parseInt(config.width),
          height: parseInt(config.height),
          deviceScaleFactor: parseInt(config.scale)
        });

        // Screen captured
        node.status({ fill: "green", shape: "dot", text: `viewport changed` });
        send(msg);
      } catch (e) {
        // If an error occured
        node.error(e);
        // Update the status
        node.status({ fill: "red", shape: "dot", text: e });
        // And update the message error property
        msg.error = e;
        send(msg);
      }

      // Clear status of the node
      setTimeout(() => {
        done();
        node.status({});
      }, (msg.error) ? 10000 : 3000);
    });
    this.on("close", function () {
      node.status({});
    });
    oneditprepare: function oneditprepare() {
      $("#node-input-name").val(this.name);
    }
  }
  RED.nodes.registerType("puppeteer-page-viewport", PuppeteerPageViewport);
};


puppeteer/page/localStorage.js
module.exports = function (RED) {
  function PuppeteerPageLocalstorage(nodeConfig) {
    RED.nodes.createNode(this, nodeConfig);
    var node = this; // Referencing the current node

    this.on("input", async function (msg, send, done) {
      try {
        await msg.puppeteer.page.evaluate((c) => {
          window.localStorage.setItem(c.key, c.value);
        }, nodeConfig);
        // Sending the msg
        send(msg);

      } catch (e) {
        // If an error occured
        node.error(e);
        // Update the status
        node.status({ fill: "red", shape: "dot", text: e });
        // And update the message error property
        msg.error = e;
        send(msg);
      }

      // Clear status of the node
      setTimeout(() => {
        done();
        node.status({});
      }, (msg.error) ? 10000 : 3000);
    });
    this.on("close", function () {
      node.status({});
    });
    oneditprepare: function oneditprepare() {
      $("#node-input-name").val(this.name);
      $("#node-input-waitUntil").val(this.waitUntil);
    }
  }
  RED.nodes.registerType("puppeteer-page-localstorage", PuppeteerPageLocalstorage);
};


puppeteer/browser/launch.js
const puppeteer = require("puppeteer");

module.exports = function (RED) {
  function PuppeteerBrowserLaunch(nodeConfig) {
    RED.nodes.createNode(this, nodeConfig);
    nodeConfig.defaultViewport = null; // Setting the node's default viewport
    nodeConfig.ignoreHTTPSErrors = true; // Setting the node's ignoreHttpsErrors property
    const node = this; // Referencing the current node

    this.on("input", async function (msg, send, done) {
      const args = msg.args || [];

      try {
        node.status({fill: "blue", shape: "dot", text: "Launching..."});
        // If debugport is specified
        if (nodeConfig.debugport != 0) {
          try {
            // Trying to connect to already existing
            // browser with node's config
            msg.puppeteer = {
              browser: await puppeteer.connect({
                ...nodeConfig,
                browserURL: `http://127.0.0.1:${nodeConfig.debugport}`,
              }),
            };

            node.status({
              fill: "green",
              shape: "dot",
              text: "Attached to existing browser",
            });
          } catch (e) { // If there is no existing browser
            node.status({
              fill: "gray",
              shape: "dot",
              text: "No existing browser detected, launching new one...",
            });

            // Launch a new browser with node's config
            msg.puppeteer = {
              browser: await puppeteer.launch({
                ...nodeConfig,
                args: [...args, `--remote-debugging-port=${nodeConfig.debugport}`],
              }),
            };
            // Browser launched sucessfully
            node.status({fill: "green", shape: "dot", text: "Launched"});
          }
        }
        // If there is no existing browser
        // or rather the puppeteer message property is undefined
        if (msg.puppeteer == undefined) {
          // Launch a new browser with node's config
          msg.puppeteer = {
            browser: await puppeteer.launch({
              ...nodeConfig,
              args: [...args, `--remote-debugging-port=${nodeConfig.debugport}`],
            }),
          };
          // Browser launched sucessfully
          node.status({fill: "green", shape: "dot", text: "Launched"});
        }
        // Get the page and set it to the puppeteer property of msg
        msg.puppeteer.page = (await msg.puppeteer.browser.pages())[0];
        msg.puppeteer.page.setDefaultTimeout(nodeConfig.timeout);

        // Get the cookies from input
        let cookies =
          nodeConfig.cookies !== ""
            ? nodeConfig.cookies
            : JSON.stringify(msg.payload);
        // Parse the cookies
        try {
          cookies = JSON.parse(cookies);
        } catch (e) {
          cookies = [];
        }

        // If cookies are passed through on lauch, set them for the page object
        try {
          // Setting the cookies
          for (const cookie of cookies) {
            await msg.puppeteer.page.setCookie(cookie);
          }
        } catch (e) {
          node.status({fill: "yellow", shape: "dot", text: e});
        }
        // Sending the msg
        send(msg);

      } catch (e) {
        // If an error occured
        node.error(e);
        // Update the status
        node.status({fill: "red", shape: "dot", text: e});
        // And update the message error property
        msg.error = e;
        send(msg);
      }

      // Clear status of the node
      setTimeout(() => {
        done();
        node.status({});
      }, (msg.error) ? 10000 : 3000);
    });
    this.on("close", function () {
      node.status({});
    });
    oneditprepare: function oneditprepare() {
      $("#node-input-timeout").val(nodeConfig.timeout);
      $("#node-input-slowMo").val(nodeConfig.slowMo);
      $("#node-input-headless").val(nodeConfig.headless);
      $("#node-input-debugport").val(nodeConfig.debugport);
      $("#node-input-devtools").val(nodeConfig.devtools);
      $("#node-input-name").val(nodeConfig.name);
    }
  }

  RED.nodes.registerType("puppeteer-browser-launch", PuppeteerBrowserLaunch);
};
